rest_framework_json_api/views.py

```diff
@@ -62,10 +70,52 @@ class MyViewSet(viewsets.ModelViewSet):
         return qs
 
 
+class SelectAndPrefetchForIncludesMixin(object):
+    """
+    This mixin provides a helper attributes to select or prefetch related models
+    based on the include specified in the URL.
+
+    __all__ can be used to specify a prefetch which should be done regardless of the include
+
+    .. code:: python
+
+    # When MyViewSet is called with ?include=author it will prefetch author and authorbio
+    class MyViewSet(viewsets.ModelViewSet):
+        queryset = Book.objects.all()
+        prefetch_for_includes = {
+            '__all__': [],
+            'category.section': ['category']
+        }
+        select_for_includes = {
+            '__all__': [],
+            'author': ['author', 'author__authorbio'],
+        }
+    """
+    def get_queryset(self):
+        qs = super(SelectAndPrefetchForIncludesMixin, self).get_queryset()
+
+        includes = self.request.GET.get('include', '').split(',') + ['__all__']
+
+        if hasattr(self, 'select_for_includes'):
+            selects = [self.select_for_includes.get(inc) for inc in includes]
+            qs = qs.select_related(*selects)
+
+        if hasattr(self, 'prefetch_for_includes'):
+            prefetches = [self.prefetch_for_includes.get(inc) for inc in includes]
+            qs = qs.prefetch_related(*prefetches)
+
+        return qs
+
+
 class AutoPrefetchMixin(object):
```

Comment:
I would also add a new `AutoPreloadMixin` for naming consistenency. It should also automatically auto load select_related (as outline in #591)

Reply:
Well, that might cover not all cases. What if I have related model 
```python
class SomeModel():
    product = ForeignKey('')
```

then in serializers.py I have:
```python 
class SomeSerializer():
    product_info = RelatedField(source='product')
```

That Mixin will try to select_related `product_info`  which does not exist, right ? But we can easily handle that with customer dictionary like we do with `prefetch_for_includes`

What do you think ?

Reply:
`select_for_includes` dictionary certainly makes sense. What I would like to see is for the common cases that auto loading happens e.g. when source of related field references a foreign key or a reverse relation.
If it doesn't (e.g. in case of using SerializerMethodResourceRelated field or similar) user can still use dictionary `select_for_includes`.

This way this PR it is consistent with prefetching where it also auto prefetches but it is also possible to manually prefetch if necessary.

Reply:
So we just need just a check like below?
```python
class AutoPrefetchMixin(object):
    def get_queryset(self, *args, **kwargs):
        """ This mixin adds automatic prefetching for OneToOne and ManyToMany fields. """
        qs = super(AutoPrefetchMixin, self).get_queryset(*args, **kwargs)

        if hasattr(self, 'select_for_includes'): 
            return qs

        included_resources = get_included_resources(self.request)
        ......
```

Reply:
Not quite what I wanted to say... :wink:

Currently the `AutoPrefetchMixin` only uses prefetch_related [see code](https://github.com/django-json-api/django-rest-framework-json-api/blob/master/rest_framework_json_api/views.py#L107).

As this PR completes manual loading by `select_related` the auto loading should also for completeness and better performance distinguish between `select_related` and `prefetch_related`.

Meaning if it is a to one relation use `select_related` or otherwise `prefetch_related` as today.

Hope this makes sense?

Reply:
Yes, it does. Sorry for late response.
I'm not sure we can handle it this way.
Example:

```python
class Order():
    customer = ForeignKey()

class Customer():
    preferences = ManyToManyField()


class Preference()
    name = CharField()
```
Let's say we have request `GET api/orders/?include=customer__preferences`, so it should produce queryset like `Order.objects.select_related(customer__preferences)` which will fail, because `preference` should be FK but not m2m relation.
And we can't do `select_related` and pass custom queryset like this:
Order.objects.select_related(Select('customer', queryset=Customer.objects.prefetch_related('preferences')))

So, `prefetch_related` covers _all_ cases. For everything else use `PreloadIncludesMixin`

Let me know if you know any other way we can prefetch and select automatically

Reply:
It needs to be a mix of course what performs better (select related if possible of prefetch_related if M2M)
Anyhow on second thought it is properly better to work on improving prefetch related in another PR at a later point (e.g. also in cases where include is not used but prefetching would be necessary for creating relationships ids)

So let's keep the changes to what there are at the moment. Could you follow up on the other points in the checklist (test, changelog, documenation etc.)? That would be great.

Reply:
@sliverc I've added a case when we definitely can use `select_related`. Let's say we have a relation chain: `OrderItem -> Order -> Customer`. They're all "[toOne](https://github.com/django-json-api/django-rest-framework-json-api/pull/600/files#diff-0ca988ea2e19e056dcd62de1c07683ddR142)" descriptors, so we can use select_related for them. Otherwise - use prefetch_related.

Also, I've merged these two mixins (AutoPreloadMixin and PrefetchForIncludesHelperMixin) into one, so it preloads automatically, but user/developer might override specific include(s) via `select_for_includes` and/or `prefetch_for_includes`. PrefetchForIncludesHelperMixin is no longer a base class of ModelViewSet, but still in the code as deprecated,  will remove it after this PR will be released.

Test added, docs updated.

Pls take another look.
Thank you